### PR TITLE
Ordered k8s apply

### PR DESCRIPTION
**Pull request by moosecanswim**
OPS-58

### CODE REVIEW

File: pkg/mesh_install/install.go [hunk 1]
```diff
@@ -155,3 +157,80 @@ func (i *Installer) triggerInjectedSidecarReconciliation() {
 		}
 	}
 }
+
+type KindSortOrder []string
+
+var InstallOrder KindSortOrder = []string{
```
Comment (alecholmez) [inline pkg/mesh_install/install.go:163]: You've created a type mask so you don't need to redeclare in the literal syntax
```suggestion
var InstallOrder = KindSortOrder{
```

File: pkg/mesh_install/install.go [hunk 2]
```diff
@@ -155,3 +157,80 @@ func (i *Installer) triggerInjectedSidecarReconciliation() {
 		}
 	}
 }
+
+type KindSortOrder []string
```
Comment (alecholmez) [inline pkg/mesh_install/install.go:161]: Doc string?

File: pkg/mesh_install/install.go (hunk 1 above)
Comment (alecholmez) [reply pkg/mesh_install/install.go:163]: Also Doc string since it's exported

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,10 @@
 
 ## vNEXT
 
+### Changed
```
Comment (alecholmez) [inline CHANGELOG.md:5]: Should this be an add? The comment notes an `Added` feature. I'd reword to match the semantics of the changelog header

File: pkg/mesh_install/install.go [hunk 4]
```diff
@@ -155,3 +157,80 @@ func (i *Installer) triggerInjectedSidecarReconciliation() {
...
+	"Job",
+	"CronJob",
+	"IngressClass",
+	"Ingress",
+	"APIService",
+}
+
+func sortManifestsByKind(manifests []client.Object, ordering KindSortOrder) []client.Object {
+	sort.SliceStable(manifests, func(i, j int) bool {
+		return lessByKind(manifests[i], manifests[j], manifests[i].GetObjectKind().GroupVersionKind().Kind, manifests[j].GetObjectKind().GroupVersionKind().Kind, ordering)
```
Comment (alecholmez) [inline pkg/mesh_install/install.go:205]: This line is super long, we can break it a tiny bit?

**Review by danielpcox — APPROVED** [review]
Heck yes

File: pkg/mesh_install/install.go [hunk 5]
```diff
@@ -155,3 +157,80 @@ func (i *Installer) triggerInjectedSidecarReconciliation() {
...
+	})
+
+	return manifests
+}
+
+func lessByKind(a interface{}, b interface{}, kindA string, kindB string, o KindSortOrder) bool {
```
Comment (alecholmez) [inline pkg/mesh_install/install.go:211]: Top level scoped variables and parameters should generally be the most descriptive in Go. Short variable names are intended for tightly scoped blocks, like `err` or `i` in a tiny for loop.
```suggestion
func lessByKind(a interface{}, b interface{}, kindA string, kindB string, order KindSortOrder) bool {
```

Comment (alecholmez) [reply pkg/mesh_install/install.go:211]: Also do `a` and `b` need to be `interface{}`? Can they be `client.Object` since that's what gets passed in from the wrapper sort function?

**Review by alecholmez — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (moosecanswim) [reply CHANGELOG.md:5]: Not really a feature for users.  It modifies how k8s stuff are applied.

Comment (alecholmez) [reply CHANGELOG.md:5]: I'm looking for something like:
```
- Changed the k8s manifest installation process to specify an order of priority given the resource kind
```

**Comment by alecholmez** [comment]
Also FWIW I think we need to drop the helm license in this source code too. This is their implementation so we should cover our bases legally. Maybe it's worth breaking this out into a separate file so we can drop the helm license at the top of the file? It's apache 2 which is good just need to make sure we give credit.

File: CHANGELOG.md (hunk 3 above)
Comment (moosecanswim) [reply CHANGELOG.md:5]: Updated comment and moved into fixed section.

File: pkg/mesh_install/install.go (hunk 4 above)
Comment (moosecanswim) [reply pkg/mesh_install/install.go:205]: broke it up into more lines

File: pkg/mesh_install/install.go (hunk 5 above)
Comment (moosecanswim) [reply pkg/mesh_install/install.go:211]: didnte even need that.  just used the string kind cause thats all we were using

**Comment by moosecanswim** [comment]
Looking at this it looks like we just need a notice file that sites helm and states that we derived code from what they have.  https://www.apache.org/licenses/LICENSE-2.0.html#redistribution

**Comment by alecholmez** [comment]
> Looking at this it looks like we just need a notice file that sites helm and states that we derived code from what they have. https://www.apache.org/licenses/LICENSE-2.0.html#redistribution

Beautiful, that works for me. As long as we give credit here I'm good with that
